tests/cbmc/proofs/s2n_stuffer_skip_read_until/s2n_stuffer_skip_read_until_harness.c

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "stuffer/s2n_stuffer.h"
+
+#include <assert.h>
+#include <string.h>
+
+#include <cbmc_proof/cbmc_utils.h>
+#include <cbmc_proof/make_common_datastructures.h>
+#include <cbmc_proof/proof_allocators.h>
+
+void s2n_stuffer_skip_read_until_harness() {
+    /* Non-deterministic inputs. */
+    struct s2n_stuffer *stuffer = cbmc_allocate_s2n_stuffer();
+    __CPROVER_assume(s2n_stuffer_is_valid(stuffer));
+    __CPROVER_assume(s2n_blob_is_bounded(&stuffer->blob, MAX_BLOB_SIZE));
+    char *target = nondet_bool() ? ensure_c_str_is_allocated(MAX_STRING_LEN) : NULL;
+
+    /* Save previous state from stuffer. */
+    struct s2n_stuffer old_stuffer = *stuffer;
+    struct store_byte_from_buffer old_byte_from_stuffer;
+    save_byte_from_blob(&stuffer->blob, &old_byte_from_stuffer);
+
+    /* Operation under verification. */
+    s2n_stuffer_skip_read_until(stuffer, target);
+
+    assert(stuffer->write_cursor == old_stuffer.write_cursor);
+    assert(stuffer->high_water_mark == old_stuffer.high_water_mark);
+    assert(stuffer->alloced == old_stuffer.alloced);
+    assert(stuffer->growable == old_stuffer.growable);
+    assert_blob_equivalence(&stuffer->blob, &old_stuffer.blob, &old_byte_from_stuffer);
```

Comment:
don't we have a macro for this?

Reply:
Note in this case, because we can't guarantee that `tainted` will be equivalence and we don't have a macro to ignore that and `read_cursor`.

Reply:
It surprises me that this could taint a stuffer. @camshaft is that what we want here?

Reply:
@danielsn `s2n_stuffer_skip_read_until` uses `s2n_stuffer_raw_read`, which could taint a stuffer.

Reply:
My understanding is that it's marked tainted when you have multiple pointers into the buffer. Doing a reallocation would invalidate all of those pointers and there's not a mechanism to update their values.

In the case of `raw_read` it hands out a pointer, which requires the tainted flag to be set, as the code doesn't know how long that pointer will live. That doesn't, however, prevent us from restoring the previous value of the tainted flag after the function returns, as that pointer will be discarded.

Ideally all of the internal functions that use `raw_read` and don't hand out a pointer would have this same behavior.

Reply:
Tldr; I would expect this to not set the tainted flag.

Reply:
@camshaft in that case, right after the `raw_read` call, we should move tainted back to what it was, correct?

Reply:
I think it would be best if we did something like this:

```c
unsigned int was_tained = stuffer->tainted;
char *actual = s2n_stuffer_raw_read(stuffer, len);
notnull_check(actual);

bool is_match = strncmp(actual, target, len) == 0;

/* restore tainted state after `actual` is no longer needed */
stuffer->tainted = was_tained;

if (is_match) {
    return S2N_SUCCESS;
} else {
    /* If string doesn't match, rewind stuffer to 1 byte after last read */
    GUARD(s2n_stuffer_rewind_read(stuffer, len - 1));
    continue;
}
```

Reply:
This would still lead to the case where `actual == NULL` and we'd return a tained stuffer. So, completely replaced the `s2n_stuffer_raw_read` call. Could you take another look? @danielsn @camshaft